### ISSUE TITLE
RKE2 Provisioning: Day 2 Ops - On Demand/Recurring Etcd Snapshot and Restore with Options

### ISSUE THREAD

**Issue by deniseschannon** [rancher/rancher#34417]
This is to validate S3 Support for snapshot and restore. 

```
apiVersion: provisioning.cattle.io/v1
kind: Cluster
metadata:
  name: #string
  annotations:
    {}
#    key: string
  labels:
    {}
#    key: string
  namespace: fleet-default
spec:
  cloudCredentialSecretName: do
  kubernetesVersion: v1.21.0-alpha2+rke2r1
  rkeConfig:

#    etcdSnapshotCreate:
#      name: string
#      nodeName: string
#      s3:
#        bucket: string
#        cloudCredentialName: string
#        endpoint: string
#        endpointCA: string
#        folder: string
#        region: string
#        skipSSLVerify: boolean
```

Note: Things changed with https://github.com/rancher/rancher/issues/34088
An s3-compatible cloud credential was created, you can create one and specify when making a cluster, and override defaults (if any are set).

**Comment by Oats87** [comment]
Now that this is merged into `release/v2.6`, here is the summary of changes for this

1. `etcd snapshot` information is no longer stored on the `clusters.provisioning.cattle.io/status` field, they are now stored in dedicated `etcdsnapshot.rke.cattle.io` objects. They are in the same namespace as the cluster, and have a cluster name label (`rke.cattle.io/cluster-name`) on them that can be used to look for them.
2. `name` on the `etcdSnapshotRestore` field now refers to the name of the CR of the `etcdSnapshot` CR. `nodeName` is no longer an option as that is stored in the metadata of the `etcdSnapshot` CR.
3. The `etcdSnapshotRestore` struct has a new format. `restoreRKEConfig` is a string field that (currently) accepts three options, `none`, `kubernetesVersion`, and `all`. They are the same as what RKE1 will do when restoring from snapshot IF the metadata exists. The option should only probably be exposed to restore more than the snapshot if the snapshot `SnapshotFile.metadata` file exists. Generation behavior is the same -- it only needs to be incremented if the same snapshot is being restored more than once. The format is now:
```
type ETCDSnapshotRestore struct {
	// Name refers to the name of the associated etcdsnapshot object
	Name string `json:"name,omitempty"`

	// Changing the Generation is the only thing required to initiate a snapshot restore.
	Generation int `json:"generation,omitempty"`
	// Set to either none (or empty string), all, or kubernetesVersion
	RestoreRKEConfig string `json:"restoreRKEConfig,omitempty"`
}
```

4. A user should not be able to `delete` an `etcdsnapshot` CR manually -- if they want to delete it, they should use K3s/RKE2 to delete it on the node itself and allow the Rancher controllers to "trickle back" the deletion (and make the object disappear). Of course, this doesn't stop a user from trying to `kubectl delete etcdsnapshot` but if they do this, the object will just pop back up.
5. The `etcdsnapshot` objects names are of format `<cluster-name>-<snapshot-name>-(local|s3)`

You may be asking, "how the heck do I test this", it's really easy. 

Just use `kubectl` to `get` `etcdsnapshots.rke.cattle.io` in the namespace of your cluster (`fleet-default`), and specify it for the `name` on the `etcdSnapshotRestore` struct. If you want to test kubernetesVersion restore or full cluster spec restore, set `restoreRKEConfig` to either `all` or `kubernetesVersion`.

**Comment by slickwarren** [comment]
blocked on https://github.com/rancher/rancher/issues/36503

**Comment by sowmyav27** [comment]
Moving this to 2.6.5, as a thorough validation will be done post 2.6.4.
For 2.6.4, these tests will be covered
- validate if P0 local etcd snapshots + restore work
- validate if P0 s3 etcd snapshots + restore work
- Log a UI bug to Hide the snapshot Now and Restore actions on the cluster and on snapshots page

**Comment by Josh-Diamond** [comment]
Blocked - awaiting PR merge for `https://github.com/rancher/rancher/issues/36941`

**Comment by Josh-Diamond** [comment]
Testing blocked; awaiting UI fix for https://github.com/rancher/dashboard/issues/5553

**Comment by Josh-Diamond** [comment]
Ticket #34417 - Test Results

With Docker on single-node instance:

Veri-FAILED on rancher `v2.6-b0b517977ce06e2bab7b6d827912a0011a779da5-head`:
1. Fresh install of rancher `v2.6-head`
2. Provision a downstream RKE2 Custom cluster, with snapshots configured to use local etcd - [default]
3. Take a local etcd snapshot
4. Add a namespace to default project, `here-for-now`
5. Restore to local etcd snapshot taken in step 3
6. Verified - Local etcd snapshot restore is successful - [namespace from step 4 removed]
7. Configure cluster to use s3 bucket for snapshots and restore
8. Add a namespace to default project, `s3`
9. Take a snapshot with s3
10. Veri-FAILED - Blocked by #36941; Will resume testing when un-blocked

Screenshots:

`Local etcd`
<img width="1612" alt="Screen Shot 2022-04-12 at 2 30 40 PM" src="https://user-images.githubusercontent.com/46494969/163058887-5f366ee0-9d7c-44b0-a66d-0ea52ca97a24.png">

<img width="1618" alt="Screen Shot 2022-04-12 at 2 32 26 PM" src="https://user-images.githubusercontent.com/46494969/163058906-123dce0b-c88b-40e6-a414-3bd5135a1821.png">

<img width="1614" alt="Screen Shot 2022-04-12 at 2 37 51 PM" src="https://user-images.githubusercontent.com/46494969/163058923-3a768f8c-45b9-4e1f-b08e-87bc76d9fc26.png">

<img width="1605" alt="Screen Shot 2022-04-12 at 2 38 12 PM" src="https://user-images.githubusercontent.com/46494969/163058941-a7e8f0a8-130b-4c8d-a576-8f10df602c1b.png">


`s3 - BLOCKED`
<img width="1369" alt="Screen Shot 2022-04-12 at 3 30 35 PM" src="https://user-images.githubusercontent.com/46494969/163065356-9a8d10a1-c8e9-417f-b24f-7eceeef12685.png">

**Comment by Josh-Diamond** [comment]
# BLOCKED - by ticket [#36943](https://github.com/rancher/rancher/issues/36943)
---
Ticket #34417 - Test Results [resumed] 

With Docker on a single-node install:

Veri-FAILED on rancher `v2.6-56f0c2d81f99fab8e1c40aac0cb05ebd2c8b1aef-head`:
1. Fresh install of rancher `v2.6-head`
2. As admin, create a standard user
3. As standard user, provision a downstream RKE2 Custom cluster, with snapshots configured to use local etcd - [default]
4. Take a local etcd snapshot
5. Veri-FAILED - Standard user cannot take snapshots - Blocked by [#36943](https://github.com/rancher/rancher/issues/36943)

**Comment by Josh-Diamond** [comment]
Re-opened - Standard user unable to access s3 snapshots in RKE2 Custom cluster

---
Ticket #34417 - Test Results [resumed - pt.2]

With Docker on single-node instance:

Veri-FAILED on rancher `v2.6-6c12bee2d0bee16a2854df5471aaa0cdee5b9bf1-head`:

1. Fresh install of rancher `v2.6-head`
2. As admin, create a standard user
3. As standard user, provision a downstream RKE2 Custom cluster, with snapshots configured to use local etcd - [default]
4. Take a local etcd snapshot
5. Add a namespace to default project, `here-for-now-local`
6. Restore to local etcd snapshot taken in step 4
7. Verified - Local etcd snapshot restore is successful - [namespace from step 5 removed]
8. Configure cluster to use s3 bucket for snapshots and restore
10. Take a snapshot with s3
12. Veri-FAILED - standard user's s3 snapshots are not accessible in rancher UI, however, they are saved and stored in s3 bucket

Screenshots:

`Local - pass`
<img width="1254" alt="here-for-now-local" src="https://user-images.githubusercontent.com/46494969/163891493-e81378b8-e49f-4f15-afaa-ab5dda7f2162.png">

<img width="1260" alt="waiting-for-etcd-restore-local" src="https://user-images.githubusercontent.com/46494969/163891518-8d6119a2-cd63-4032-adc3-49cc6cde2f6a.png">

<img width="1260" alt="restored-local" src="https://user-images.githubusercontent.com/46494969/163891530-3a205d88-9068-4c7c-84f8-1e628e9e43b4.png">

<img width="1258" alt="restored-local-proof" src="https://user-images.githubusercontent.com/46494969/163891540-7b18c9d0-594c-444f-b250-f9ef4778c9ee.png">

`s3 - fail`
<img width="1261" alt="missing-snap-from-ui" src="https://user-images.githubusercontent.com/46494969/163891596-fe3c88ac-cd30-4bfa-a0df-13c5aadb3150.png">

<img width="942" alt="Screen Shot 2022-04-18 at 4 11 56 PM" src="https://user-images.githubusercontent.com/46494969/163891607-1972616c-0626-4b7a-a0ec-507d29364145.png">

**Comment by Oats87** [comment]
Investigating this, it appears the reason is because the `plansecret` controller is unable to reconcile periodic snapshot list output due to a bug in K3s/RKE2.

Long story short is that K3s/RKE2 have different CLI flags for `rke2 server` vs `rke2 etcd-snapshot`. Between versions 
K3s: `v1.21.5+k3s2` and `v1.21.6+k3s1`
RKE2: `v1.21.5+rke2r2` and `v1.21.6+rke2r1`

aliases were added to the `etcd-snapshot` subcommands to mimic the original `--etcd-s3` flag from the `server` subcommand, allowing for inheritance of the command from the config file parser. The issue with this is that once this aliasing was done, K3s and RKE2 refused to run `etcd-snapshot ls` commands with `--s3` defined on the command line, citing the error:
```
FATA[0000] Cannot use two forms of the same flag: etcd-s3 s3
```

Notably, we cannot simply blindly use the `--etcd-s3` flag either because the older versions of K3s/RKE2 refuse to run the `etcd-snapshot list` command with the error 
```
FATA[0000] flag provided but not defined: -etcd-s3
```

This schism of behavior requires changes on the v2prov planner side to resolve. The exact solution to this is to be determined.

**Comment by Josh-Diamond** [comment]
Blocked by #37401; will resume testing once PR is merged

**Comment by Josh-Diamond** [comment]
Ticket #34417 - Test Results [concluded]

Verified on rancher `v2.6-228a585bb25b99ee191f2911ea3c6fb0ce9e41f3-head`:
1. Fresh install of rancher `v2.6-head`
2. As admin, create a standard user
3. As standard user, provision a downstream RKE2 Custom cluster, with snapshots configured to use local etcd - [default]
4. Take a local etcd snapshot
5. Add a namespace to default project, `here-for-now-local-test`
6. Restore to local etcd snapshot taken in step 4
7. Verified - Local etcd snapshot restore is successful - [namespace from step 5 removed]
8. Configure cluster to use s3 bucket for snapshots and restore
9. Take a snapshot with s3
10. Add a namespace to default project, `here-for-now-s3-test`
11. Restore to s3 snapshot taken in step 9
12. Verified - s3 snapsot restore is successful - [namespace from step 10 removed]

Screenshots:

`Local - step 5`
<img width="1443" alt="local-snap-before-restore" src="https://user-images.githubusercontent.com/46494969/164303695-23d1f624-33ea-4617-9b0a-eb09c5c0163d.png">

`Local - step 6 - restoring`
<img width="1447" alt="local-snap-restoring" src="https://user-images.githubusercontent.com/46494969/164303805-fcaf810f-9c6b-4f36-8eef-553ed937cdda.png">

`Local - step 6 - restored`
<img width="1447" alt="local-snap-restored" src="https://user-images.githubusercontent.com/46494969/164303846-2076c0d8-51f4-4b27-a426-42f811c4e345.png">

`Local - step 7`
<img width="1438" alt="local-snap-restored-proof" src="https://user-images.githubusercontent.com/46494969/164303962-68e3a13e-7b8a-4593-96c4-f4cedb48e95f.png">

`s3 - step 10`
<img width="1444" alt="s3-before-restore" src="https://user-images.githubusercontent.com/46494969/164304109-55dbb441-e9dd-415d-a605-3727e2f8344a.png">

`s3 - step 11 - restoring`
<img width="1443" alt="s3-restoring" src="https://user-images.githubusercontent.com/46494969/164304161-631bd2b4-bc39-4acd-a008-f5950be454ab.png">

`s3 - step 11 - restored`
<img width="1445" alt="s3-restored" src="https://user-images.githubusercontent.com/46494969/164304204-7d994bdb-853a-41b8-85cb-f7d28d8f6566.png">

`s3 - step 12`
<img width="1439" alt="s3-proof" src="https://user-images.githubusercontent.com/46494969/164304246-2b39676b-0ba1-479c-87a3-75d1fcd26772.png">

`Snapshots - RancherUI`
![snaps-final](https://user-images.githubusercontent.com/46494969/164304606-ce9748ee-ac6e-4732-9261-430f190a3f93.png)

`Snapshots - s3 Bucket`
![s3bucket](https://user-images.githubusercontent.com/46494969/164305688-03984991-8466-4fb5-b07e-c0f65000cd09.png)

**Comment by snasovich** [comment]
### Release Notes Information
This new feature needs to be added to release notes.
Caveats:
* Snapshot/restore requires k8s 1.21.6+ so it won't work on 1.21.4 and 1.21.5 (that are available, but are rather old).
